hazelcast/src/main/java/com/hazelcast/nio/NodeIOService.java

```diff
@@ -49,12 +57,77 @@
 @PrivateApi
 public class NodeIOService implements IOService {
 
+    private static final ILogger LOGGER = Logger.getLogger(NodeIOService.class);
+
     private final Node node;
     private final NodeEngineImpl nodeEngine;
+    private final RestApiConfig restApiConfig;
+    private final MemcacheProtocolConfig memcacheProtocolConfig;
 
     public NodeIOService(Node node, NodeEngineImpl nodeEngine) {
         this.node = node;
         this.nodeEngine = nodeEngine;
+        restApiConfig = initRestApiConfig(node.getProperties(), node.getConfig());
+        memcacheProtocolConfig = initMemcacheProtocolConfig(node.getProperties(), node.getConfig());
+    }
+
+    /**
+     * Initializes {@link RestApiConfig} if not provided based on legacy group properties. Also checks (fails fast)
+     * if both the {@link RestApiConfig} and system properties are used.
+     */
+    @SuppressWarnings("deprecation")
+    private static RestApiConfig initRestApiConfig(HazelcastProperties properties, Config config) {
+        RestApiConfig restApiConfig = config.getRestApiConfig();
+        if (restApiConfig != null) {
+            // ensure the legacy Hazelcast group properties are not provided
+            ensurePropertyNotConfigured(properties, GroupProperty.REST_ENABLED);
+            ensurePropertyNotConfigured(properties, GroupProperty.HTTP_HEALTHCHECK_ENABLED);
+        } else {
+            restApiConfig = new RestApiConfig();
+            if (checkDeprecatedProperty(properties, GroupProperty.REST_ENABLED)) {
+                restApiConfig.setEnabled(true);
+                restApiConfig.enableAllGroups();
+            }
+            if (checkDeprecatedProperty(properties, GroupProperty.HTTP_HEALTHCHECK_ENABLED)) {
+                restApiConfig.setEnabled(true);
+                restApiConfig.enableGroups(RestEndpointGroup.HEALTH_CHECK);
+            }
+        }
+        return restApiConfig;
+    }
+
+    @SuppressWarnings("deprecation")
+    private static MemcacheProtocolConfig initMemcacheProtocolConfig(HazelcastProperties properties, Config config) {
+        MemcacheProtocolConfig memcacheProtocolConfig = config.getMemcacheProtocolConfig();
+        if (memcacheProtocolConfig != null) {
+            // ensure the legacy Hazelcast group property is not provided
+            ensurePropertyNotConfigured(properties, GroupProperty.MEMCACHE_ENABLED);
+        } else {
+            memcacheProtocolConfig = new MemcacheProtocolConfig();
+            if (checkDeprecatedProperty(properties, GroupProperty.MEMCACHE_ENABLED)) {
+                memcacheProtocolConfig.setEnabled(true);
+            }
+        }
+        return memcacheProtocolConfig;
+    }
+
+    private static void ensurePropertyNotConfigured(HazelcastProperties properties, HazelcastProperty hazelcastProperty)
```

Comment:
Can the `ensure...` methods live in the `ConfigValidator` class?

Reply:
You're right, I'll move the `ensurePropertyNotConfigured` and `checkDeprecatedProperty` methods to `ConfigValidator`.